# Исследование надежности заемщиков. <a id='intro'></a>

<h1>Cодержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание" data-toc-modified-id="Описание-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li></ul></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Категоризация-по-количеству-детей" data-toc-modified-id="Категоризация-по-количеству-детей-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Категоризация по количеству детей</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>Вывод 1:</a></span></li><li><span><a href="#Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>Вывод 2:</a></span></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-8.5"><span class="toc-item-num">8.5&nbsp;&nbsp;</span>Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-8.6"><span class="toc-item-num">8.6&nbsp;&nbsp;</span>Вывод 3:</a></span></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-8.7"><span class="toc-item-num">8.7&nbsp;&nbsp;</span>Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-8.8"><span class="toc-item-num">8.8&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Общий вывод:</a></span></li><li><span><a href="#Рекомендации-по-повышению-качества-данных" data-toc-modified-id="Рекомендации-по-повышению-качества-данных-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Рекомендации по повышению качества данных</a></span></li></ul></div>

## Описание

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга**.

## Обзор данных

Импортируем библиотеку `pandas`.

In [1]:
import pandas as pd

Прочитаем файл `data.csv` и сохраним его в переменную df.

In [3]:
df = pd.read_csv('data.csv')

Выведем на экран первые 5 строк таблицы.

In [4]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Получим общую информацию о таблице.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Рассмотрим полученную информацию подробнее.

В датафрейме 12 столбцов из которых 4 столбца с количественными, 8 с категориальными значениями. 
Массив содержит следующие типы данных — `object`, `int64`, `float64`.

Подробно разберём, какие в переменной **df** столбцы и какую информацию они содержат:
* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;
* `debt` — имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита.

Каждая строка таблицы содержит данные о клиентах банка. Имеющиеся данные позволяют нам судить: 
- об уровне образования клиентов банка;
- о наличии детей у клиентов; 
- о семейном положении клиентов;
- о стаже работы клиентов;
- о целях получения кредита клиентами. 

**Выводы**

Выявленные аномалии, пропуски в данных:

- Количество строк в датафрейме 21 525, каждая из которой содержит данные о клиентах банка. Однако, количество не пустых строк в столбце с общим трудовым стажем `days_employed` и ежемесячным доходом `total_income` меньше на 10% и составляет 19 351. Объем пропусков не маленький, удаление всех строк, в которых встречаются пропуски в указанных столбцах или заполнение их значениями по умолчанию, может отразиться на точности исследования;
- В столбце с возрастами клиентов банка `dob_years` встречаются нулевые значения. Данные значения нам пригодятся для категоризации клиентов по возрасту;
- Столбец с информацией о стажах работы клиентов `days_employed` содержит отрицательные значения, стаж работы клиентов может быть нулевым, но не отрицательным. Кроме того, в столбце содержатся данные о клиентах, имеющих опыт работы свыше 100, что в принципе также сложно представить;
- В столбце с количеством детей у клиентов `children` содержатся значения равные -1 и 20, иначе говоря, у банка есть клиенты у которых количество детей -1 и 20. Одназначно можно сказать, что количество детей может быть нулевым, но не отрицательным, наряду с этим наличие клиентов с 20-ью детьми также кажется недостоверной информацией, поскольку такие клиенты выпадают из общей вариации количества детей клиентов банка;
- В столбце с информацией о половой принадлежности клиентов банка `gender` встречается значение "XNA", принадлежность которой к имеющимся полам сложно установить;
- В столбце с информацией об уровнях образования клиентов `education` наблюдается разнорегисторная запись, например, значения "Среднее" и "среднее" определены, как 2 разных значения, что неверно. Также в столбце с информацией о семейных статусах не учтен нижний регистр для семейного статуса "Не женат / не замужем", что не критично;


Для проверки рабочих гипотез особенно ценны столбцы: `children`, `family_status`, `debt`, `total_income`, `purpose`. В некоторых из перечисленных столбцов существуют большие пропуски.




[В начало](#intro).

## Предобработка данных

### Заполнение пропусков

Прежде, чем приступить к заполнению пропусков, подкорректируем наименования столбцов для удобства исследования.

Переименуем следующие столбцы:

`children` на `children_count`;

`dob_years` на `age`;

`education` на `education_level`;

`income_type` на `employee_type`;

`total_income` на `monthly_income`.

In [5]:
df = df.rename(columns={'children': 'children_count', 
                        'dob_years': 'age', 
                        'education': 'education_level', 
                        'income_type': 'employee_type', 
                        'total_income': 'monthly_income'})
df.head()

children_count  days_employed  age education_level  education_id  \
0               1   -8437.673028   42          высшее             0   
1               1   -4024.803754   36         среднее             1   
2               0   -5623.422610   33         Среднее             1   
3               3   -4124.747207   32         среднее             1   
4               0  340266.072047   53         среднее             1   

      family_status  family_status_id gender employee_type  debt  \
0   женат / замужем                 0      F     сотрудник     0   
1   женат / замужем                 0      F     сотрудник     0   
2   женат / замужем                 0      M     сотрудник     0   
3   женат / замужем                 0      M     сотрудник     0   
4  гражданский брак                 1      F     пенсионер     0   

   monthly_income                     purpose  
0   253875.639453               покупка жилья  
1   112080.014102     приобретение автомобиля  
2   145885.952297               покупка жилья  
3   267628.550329  дополнительное образование  
4   158616.077870             сыграть свадьбу

Использование медианного значения, найденного на основании всего столбца с общим доходом может повышать или понижать уровень дохода для отдельных типов занятости, т.е. не будем использовать "среднюю температуру по всей больнице". Заполним пропущенные значения в столбце с ежемесячным доходом `monthly_income` медианным значением по каждому типу занятости по столбцу `employee_type`.

Посчитаем количество пропусков по каждому типу занятости.

In [6]:
df.loc[df['monthly_income'].isna(), 'employee_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: employee_type, dtype: int64

Найдем медианные значения уровня дохода для каждого типа занятости.

In [7]:
employee_types_medians = df.groupby('employee_type')['monthly_income'].median()
employee_types_medians

employee_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: monthly_income, dtype: float64

Заполним столбец c ежемесячным доходом найденными медианными значениями для каждого отдельного типа занятости, в которых нам встретились пропуски.

In [8]:
df.loc[(df['monthly_income'].isna()) & (df['employee_type'] == 'сотрудник'), 'monthly_income'] = employee_types_medians[6]
df.loc[(df['monthly_income'].isna()) & (df['employee_type'] == 'компаньон'), 'monthly_income'] = employee_types_medians[3]
df.loc[(df['monthly_income'].isna()) & (df['employee_type'] == 'пенсионер'), 'monthly_income'] = employee_types_medians[4]
df.loc[(df['monthly_income'].isna()) & (df['employee_type'] == 'госслужащий'), 'monthly_income'] = employee_types_medians[2]
df.loc[(df['monthly_income'].isna()) & (df['employee_type'] == 'предприниматель'), 'monthly_income'] = employee_types_medians[5]

Проверим все ли пропуски в столбце с ежемесячным доходом были заполнены.

In [9]:
df['monthly_income'].isna().sum()

0

В столбце с информацией о возрасте клиентов банка встречаются нули.

In [10]:
df.loc[df['age'] == 0, 'employee_type'].value_counts()

сотрудник      55
пенсионер      20
компаньон      20
госслужащий     6
Name: employee_type, dtype: int64

Найдем медианные значения по возрасту клиентов в зависимости от типа занятости.

In [11]:
age_medians = df.groupby('employee_type')['age'].median()
age_medians

employee_type
безработный        38.0
в декрете          39.0
госслужащий        40.0
компаньон          39.0
пенсионер          60.0
предприниматель    42.5
сотрудник          39.0
студент            22.0
Name: age, dtype: float64

Заполним имеющие нулевые значения в возрастах клиентов банка медианными значениями по разным типам занятости.

In [12]:
df.loc[(df['age'] == 0) & (df['employee_type'] == 'сотрудник'), 'age'] = age_medians[6]
df.loc[(df['age'] == 0) & (df['employee_type'] == 'пенсионер'), 'age'] = age_medians[4]
df.loc[(df['age'] == 0) & (df['employee_type'] == 'компаньон'), 'age'] = age_medians[3]
df.loc[(df['age'] == 0) & (df['employee_type'] == 'госслужащий'), 'age'] = age_medians[2]

Проверим остались ли строки в столбце с нулевым возрастом клиентов.

In [13]:
df.loc[df['age'] == 0, 'age'].value_counts()

Series([], Name: age, dtype: int64)

Использование медианного значения, найденного по типу занятости при заполнении пропусков по стажу работы `days_employed` может не отражать действительную картину, поэтому заполним пропуски в стажах работы клиентов банка медианными значениями по возрасту. Для этого нам потребуется выделить возрастные категории клиентов: в качестве границ категорий выберем квантили (25%, 50% и 75%).

In [14]:
df['age'].describe()[4:7]

25%    34.0
50%    43.0
75%    53.0
Name: age, dtype: float64

Далее создадим функцию которая на входе принимает информацию о возрасте клиента и возвращает категорию. Внутри функции используем конструкцию, которая сообщит нам при наличии не валидных значений по столбцу.

In [15]:
def set_age(data):
    try:
        if data < 34:
            return 'до 34'
        if data < 43:
            return 'от 34 до 43'
        if data < 54:
            return 'от 43 до 54'
        return 'свыше 54'
    except:
        print(f"Значение {data} не попадает ни под одно условие")

Создадим новый столбец `age_frame`, который будет отражать информацию о категоризации клиентов по возрастным рамкам и заполним его нашей функцией. 

In [16]:
df['age_frame'] = df['age'].apply(set_age)

Теперь определим медианные значения для каждой возрастной группы.

In [17]:
days_employed_medians = df.groupby('age_frame')['days_employed'].median()
days_employed_medians

age_frame
до 34           -1130.581631
от 34 до 43     -1759.889098
от 43 до 54     -1839.038911
свыше 54       343033.965497
Name: days_employed, dtype: float64

Заполним пропущенные значения в столбце со стажем работы найденным медианным значением.

In [18]:
df.loc[(df['days_employed'].isna()) & (df['age_frame'] == 'до 34'), 'days_employed'] = days_employed_medians[0]
df.loc[(df['days_employed'].isna()) & (df['age_frame'] == 'от 34 до 43'), 'days_employed'] = days_employed_medians[1]
df.loc[(df['days_employed'].isna()) & (df['age_frame'] == 'от 43 до 54'), 'days_employed'] = days_employed_medians[2]
df.loc[(df['days_employed'].isna()) & (df['age_frame'] == 'свыше 54'), 'days_employed'] = days_employed_medians[3]

Проверим заполнились ли все строки в датафрейме.

In [19]:
df.isna().sum()

children_count      0
days_employed       0
age                 0
education_level     0
education_id        0
family_status       0
family_status_id    0
gender              0
employee_type       0
debt                0
monthly_income      0
purpose             0
age_frame           0
dtype: int64

**Выводы:**

- Были обнаружены пропуски в столбце с информацией о ежемесячном доходе клиентов банка и заполнены медианными значениями в зависимости от типа занятости;
- Были обнаружены нули в столбце, содержающем информацию о возрасте клиентов банка и заменены медианными значениями в зависимости от типа занятости;
- Были обнаружены пропуски в столбе с информацией о трудовом стаже клиентов и заполнены медианными значениями в зависимости от возрастной категории клиентов банка.

[В начало](#intro).

### Проверка данных на аномалии и исправления.

Данные в столбце со стажем работы клиентов банка представленны с минусом.

In [20]:
df['days_employed'].head(10)

0     -8437.673028
1     -4024.803754
2     -5623.422610
3     -4124.747207
4    340266.072047
5      -926.185831
6     -2879.202052
7      -152.779569
8     -6929.865299
9     -2188.756445
Name: days_employed, dtype: float64

Исправим найденную аномалию.

In [21]:
df['days_employed'] = abs(df['days_employed'])

Проверим устранение аномалии, повторно выведем на экран столбец со стажем работы.

In [22]:
df['days_employed'].head(10)

0      8437.673028
1      4024.803754
2      5623.422610
3      4124.747207
4    340266.072047
5       926.185831
6      2879.202052
7       152.779569
8      6929.865299
9      2188.756445
Name: days_employed, dtype: float64

Наблюдаются также аномалии с количеством детей у клиентов. У большинста клиентов количество детей варьируется от 0 до 4. Однако, среди данных представлены клиенты с количеством детей -1 и 20. 

In [23]:
df['children_count'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children_count, dtype: int64

Заменим выявленные значения -1 и 20, выпадающие из общей выборки исходя из предположения, что под значением -1 подразумевалось наличие у клиента 1-го ребенка, а под 20 - 2-ух детей.

In [24]:
df.loc[df['children_count'] == -1, 'children_count'] = 1
df.loc[df['children_count'] == 20, 'children_count'] = 2

Проверим устранены ли выявленные аномалии.

In [25]:
df['children_count'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children_count, dtype: int64

In [26]:
count = -1
for value in df['children_count'].value_counts():
    count += 1
    print(f"Доля клиентов с количеством детей равным {count} составляет {round(value/df['children_count'].value_counts().sum()*100, 2)}%")

Доля клиентов с количеством детей равным 0 составляет 65.73%
Доля клиентов с количеством детей равным 1 составляет 22.6%
Доля клиентов с количеством детей равным 2 составляет 9.9%
Доля клиентов с количеством детей равным 3 составляет 1.53%
Доля клиентов с количеством детей равным 4 составляет 0.19%
Доля клиентов с количеством детей равным 5 составляет 0.04%


Как мы видим у порядка 66% клиентов банка дети отсутствуют, около четверти клиентов имеет только одного ребенка и всего около 10% клиентов банка имеют от двух детей в семье. 

В столбце с информацей о гендерах клиентов банка обнаружена аномалия.

In [27]:
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Гендер "XNA" встречается всего раз и поскольку представителей женского пола в нашей выборке вдвое больше, чем мужского пола, то вероятность того, что данный клиент относится к женскому полу объективно больше, поэтому заменим "XNA" на "F".

In [28]:
df.loc[df['gender'] == 'XNA', 'gender'] = 'F'

Проверим был ли устранен артефакт в столбце с информацией о гендере клиентов.

In [29]:
df['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

В столбце `education_level` наюблюдаются разные формы записей уровней образования, которые могли возникнуть при внесении информации во время оформления заявки на кредит ввиду отсутствия унифицированных форм записей.

Посчитаем уникальные значения уровней образования клиентов банка.

In [30]:
df['education_level'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education_level, dtype: int64

Перед поиском дубликатов приведем все данные по образованию клиентов к нижнему регистру. В противном случае, при поиске дубликатов значения "Среднее" и "среднее" будут определены, как 2 разных значения, что неверно. 

In [31]:
df['education_level'] = df['education_level'].str.lower()

Проверим все ли значения уровней образования были приведены к нижнему регистру.

In [32]:
df['education_level'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education_level, dtype: int64

Также не все зачения столбца с семейными статусами приведены к нижнему ристру.

In [33]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [34]:
df['family_status'] = df['family_status'].str.lower()

Проверим все ли значения столбца были приведены к нижнему регистру.

In [35]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

**Выводы:**

- Были выявлены и исправлены отрицательные значения для столбца с ежемесячным доходом клиентов;
- Были выявлены и исправлены значения -1 и 20 для столбца с информацей о количестве детей клиентов банка;
- В столбце с информацией о стаже работы клиентов содержатся данные о клиентах, имеющих опыт работы свыше 100, на результат проверки гипотез по исследованию указанная аномалия никак не влияет, однако, стоит разобраться в причинах наличия таких недостоверных данных.
- Было выявлено и устранено значение "XNA" в столбце с информацией о половой принадлежности клиентов банка;
- В столбце с уровнем образования, а также в столбце с информацией о семейных статусах все значения приведены к нижнему регистру.

[В начало](#intro).

### Удаление дубликатов.

Посчитаем количество дубликатов в датафрейме.

In [36]:
df.duplicated().sum()

71

Удалим выявленные дубликаты в данных.

In [37]:
df = df.drop_duplicates().reset_index(drop=True)

Проверим были ли устранены дубликаты в данных.

In [38]:
df.duplicated().sum()

0

### Изменение типов данных.

Рассмотрим имеющиеся типы данных в датафрейме.

In [39]:
df.dtypes

children_count        int64
days_employed       float64
age                   int64
education_level      object
education_id          int64
family_status        object
family_status_id      int64
gender               object
employee_type        object
debt                  int64
monthly_income      float64
purpose              object
age_frame            object
dtype: object

В столбцах с общим стажем клиентов и ежемесячным доходом данные представлены в виде вещественных чисел.

Заменим типы данных в указанных столбцах на целые числа.

In [40]:
df['days_employed'] = df['days_employed'].astype('int')
df['monthly_income'] = df['monthly_income'].astype('int')

Проверим применились ли изменения к типам данных.

In [41]:
df.dtypes

children_count       int64
days_employed        int64
age                  int64
education_level     object
education_id         int64
family_status       object
family_status_id     int64
gender              object
employee_type       object
debt                 int64
monthly_income       int64
purpose             object
age_frame           object
dtype: object

[В начало](#intro).

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Для удобства работы с информацией, а также оптимизации данных в датафрейме создадим словари.

In [42]:
education_dict = df[['education_level', 'education_id']]
family_status_dict = df[['family_status', 'family_status_id']]

Удалим дубликаты из словарей.

In [43]:
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)

In [44]:
education_dict

education_level  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [45]:
family_status_dict

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

Удалим из исходного датафрейма столбцы с уровнем образования `education_level` и семейным статусом `family_status`.

In [46]:
df = df[['children_count', 
         'days_employed', 
         'age', 
         'education_id', 
         'family_status_id', 
         'gender', 
         'employee_type', 
         'debt', 
         'monthly_income', 
         'purpose', 
         'age_frame']]
df.head()

children_count  days_employed  age  education_id  family_status_id gender  \
0               1           8437   42             0                 0      F   
1               1           4024   36             1                 0      F   
2               0           5623   33             1                 0      M   
3               3           4124   32             1                 0      M   
4               0         340266   53             1                 1      F   

  employee_type  debt  monthly_income                     purpose    age_frame  
0     сотрудник     0          253875               покупка жилья  от 34 до 43  
1     сотрудник     0          112080     приобретение автомобиля  от 34 до 43  
2     сотрудник     0          145885               покупка жилья        до 34  
3     сотрудник     0          267628  дополнительное образование        до 34  
4     пенсионер     0          158616             сыграть свадьбу  от 43 до 54

 [В начало](#intro).

## Категоризация дохода.

Далее присвоим уровню доходов разные категории.

Для этого сначала создадим функцию, которая на входе принимает ежемесячный доход клиентов и в зависимости от уровня такого дохода присваивает ему определенную категорию.

In [47]:
def set_income_category (data):
    try:
        if data < 30000:
            return 'E'
        if data < 50000:
            return 'D'
        if data < 200000:
            return 'C'
        if data < 1000000:
            return 'B'
        return 'A'
    except:
        print(f"Значение {data} не попадает ни под одно условие")

Далее создадим новый столбец в датафрейме с отнесением доходов клиентов к указанным категориям.

In [48]:
df['monthly_income_category'] = df['monthly_income'].apply(set_income_category)
df.head()

children_count  days_employed  age  education_id  family_status_id gender  \
0               1           8437   42             0                 0      F   
1               1           4024   36             1                 0      F   
2               0           5623   33             1                 0      M   
3               3           4124   32             1                 0      M   
4               0         340266   53             1                 1      F   

  employee_type  debt  monthly_income                     purpose  \
0     сотрудник     0          253875               покупка жилья   
1     сотрудник     0          112080     приобретение автомобиля   
2     сотрудник     0          145885               покупка жилья   
3     сотрудник     0          267628  дополнительное образование   
4     пенсионер     0          158616             сыграть свадьбу   

     age_frame monthly_income_category  
0  от 34 до 43                       B  
1  от 34 до 43                       C  
2        до 34                       C  
3        до 34                       B  
4  от 43 до 54                       C

[В начало](#intro).

## Категоризация целей кредита.

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:
- операции с автомобилем;
- операции с недвижимостью;
- проведение свадьбы;
- получение образования.

Для этого сначала изучим имеющиеся данные о целях получения кредита.

In [49]:
df['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Создадим функцию, которая на входе принимает имеющиеся цели получения кредита из столбца `purpose` и обобщает их на 4 крупные категории.

In [50]:
def set_purpose_category (data):
    try:
        if data == 'на покупку своего автомобиля':
            return 'операции с автомобилем'
        if data == 'автомобиль':
            return 'операции с автомобилем'
        if data == 'сделка с подержанным автомобилем':
            return 'операции с автомобилем'
        if data == 'свой автомобиль':
            return 'операции с автомобилем'
        if data == 'на покупку подержанного автомобиля':
            return 'операции с автомобилем'
        if data == 'автомобили':
            return 'операции с автомобилем'
        if data == 'на покупку автомобиля':
            return 'операции с автомобилем'
        if data == 'приобретение автомобиля':
            return 'операции с автомобилем'
        if data == 'сделка с автомобилем':
            return 'операции с автомобилем'

        if data == 'свадьба':
            return 'проведение свадьбы'
        if data == 'на проведение свадьбы':
            return 'проведение свадьбы'
        if data == 'сыграть свадьбу':
            return 'проведение свадьбы'

        if data == 'заняться высшим образованием':
            return 'получение образования'
        if data == 'дополнительное образование':
            return 'получение образования'
        if data == 'высшее образование':
            return 'получение образования'
        if data == 'образование':
            return 'получение образования'
        if data == 'получение дополнительного образования':
            return 'получение образования'
        if data == 'получение образования':
            return 'получение образования'
        if data == 'профильное образование':
            return 'получение образования'
        if data == 'получение высшего образования':
            return 'получение образования'
        if data == 'заняться образованием':
            return 'получение образования'
        return 'операции с недвижимостью'
    except:
        print(f"Значение {data} не попадает ни под одно условие")

Создадим столбец `purpose_category` и заполним его нашей функцией.

In [51]:
df['purpose_category'] = df['purpose'].apply(set_purpose_category)
df.head()

children_count  days_employed  age  education_id  family_status_id gender  \
0               1           8437   42             0                 0      F   
1               1           4024   36             1                 0      F   
2               0           5623   33             1                 0      M   
3               3           4124   32             1                 0      M   
4               0         340266   53             1                 1      F   

  employee_type  debt  monthly_income                     purpose  \
0     сотрудник     0          253875               покупка жилья   
1     сотрудник     0          112080     приобретение автомобиля   
2     сотрудник     0          145885               покупка жилья   
3     сотрудник     0          267628  дополнительное образование   
4     пенсионер     0          158616             сыграть свадьбу   

     age_frame monthly_income_category          purpose_category  
0  от 34 до 43                       B  операции с недвижимостью  
1  от 34 до 43                       C    операции с автомобилем  
2        до 34                       C  операции с недвижимостью  
3        до 34                       B     получение образования  
4  от 43 до 54                       C        проведение свадьбы

[В начало](#intro).

## Категоризация по количеству детей

Также установим категории в зависимости от количества детей у клиентов.

Создадим функцию, которая на входе принимает информацию о количестве детей у клиента и возвращает категорию.

In [52]:
def set_children_category(data):
    try:
        if data == 0:
            return 'нет детей'
        if data == 1:
            return '1 ребенок'
        if data == 2:
            return '2 ребенка'
        return 'много детей'
    except:
        print(f"Значение {data} не попадает ни под одно условие")

In [53]:
df['children_category'] = df['children_count'].apply(set_children_category)
df.head()

children_count  days_employed  age  education_id  family_status_id gender  \
0               1           8437   42             0                 0      F   
1               1           4024   36             1                 0      F   
2               0           5623   33             1                 0      M   
3               3           4124   32             1                 0      M   
4               0         340266   53             1                 1      F   

  employee_type  debt  monthly_income                     purpose  \
0     сотрудник     0          253875               покупка жилья   
1     сотрудник     0          112080     приобретение автомобиля   
2     сотрудник     0          145885               покупка жилья   
3     сотрудник     0          267628  дополнительное образование   
4     пенсионер     0          158616             сыграть свадьбу   

     age_frame monthly_income_category          purpose_category  \
0  от 34 до 43                       B  операции с недвижимостью   
1  от 34 до 43                       C    операции с автомобилем   
2        до 34                       C  операции с недвижимостью   
3        до 34                       B     получение образования   
4  от 43 до 54                       C        проведение свадьбы   

  children_category  
0         1 ребенок  
1         1 ребенок  
2         нет детей  
3       много детей  
4         нет детей

[В начало](#intro).

## Ответы на вопросы.

### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Создадим функцию, которая будет принимать на входе столбец с количественными данными и будет возращать долю количественных данных в столбце, функция прогодится для проверки всех гипотез.

In [54]:
def make_proportion(data):
    return str(round((data.sum() / data.count()) * 100, 2)) + '%'

Создадим сводную таблицу в которой сгруппируем данные по количеству детей, найдем сумму ячеек с долгами и общее количество клиентов с числом детей от 0 до 5, а также найдем долю клиентов с количеством детей от 0 до 5 имеющих долги.

In [55]:
data_pivot = df.pivot_table(index='children_category', values='debt', aggfunc=['sum', 'count', make_proportion])
data_pivot = data_pivot.sort_values(by=('make_proportion', 'debt'))
data_pivot

sum  count make_proportion
                   debt   debt            debt
children_category                             
нет детей          1063  14091           7.54%
много детей          31    380           8.16%
1 ребенок           445   4855           9.17%
2 ребенка           202   2128           9.49%

### Вывод 1:

Данные показывают, что чем больше детей, тем выше вероятность возникновения задолженности клиентов. Однако, разница в вероятностях возникновения задолженности для клиентов с числом детей от 0 до 5 не такая большая. Полезной информацией в ответе на данный вопрос послужило бы наличие данных о сумме выданных кредитов, но, к сожалению, данная информация отсутствует.

[В начало](#intro).

### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

Вернем в новый датафрейм ранее удаленные расшифровки семейных статусов.

In [56]:
df_with_family_status = df.merge(family_status_dict, on='family_status_id', how='left')

Создадим сводную таблицу, в которой сгруппируем данные по семейному статусу, найдем сумму ячеек с долгами и общее количество клиентов с числом детей от 0 до 5, а также найдем долю клиентов с количеством детей от 0 до 5 имеющих долги.

In [57]:
data_pivot = df_with_family_status.pivot_table(index='family_status', values='debt', aggfunc=['sum', 'count', make_proportion])
data_pivot = data_pivot.sort_values(by=('make_proportion', 'debt'))
data_pivot

sum  count make_proportion
                      debt   debt            debt
family_status                                    
вдовец / вдова          63    959           6.57%
в разводе               85   1195           7.11%
женат / замужем        931  12339           7.55%
гражданский брак       388   4151           9.35%
не женат / не замужем  274   2810           9.75%

Любопытным является факт, что люди с семейными статусами **вдовец / вдова** и **в разводе** менее всего подвержены риску возникновения задолженности. В то же время люди с семейными статусами **не женат / не замужем** и **гражданский брак** более всего подвержены риску возникновения задолженности по кредиту. Переходным семейным статусом является **женат / замужем**, клиенты с таким статусом больше подвержены риску возникновения задолженности, чем клиенты с семейными статусами **вдовец / вдова** и **в разводе**, но меньше клиентов с семейными статусами **не женат / не замужем**.

Возможно, это связано с тем, что большинство клиентов с семейными статусами  **вдовец / вдова** и **в разводе** старше большинства клиентов с семейными статусами **не женат / не замужем** и **гражданский брак** и уже реализовались в жизни и не имеют проблем с доходом.

Проверим выдвинутое утверждение. Для этого рассмотрим сначала распределение клиенов по семейным статусам в зависимости от возрастной категории.

In [58]:
data_pivot = df_with_family_status.pivot_table(index='family_status', columns='age_frame', values='purpose', aggfunc=['count'])
data_pivot = data_pivot.sort_values(by=('count', 'свыше 54'))
data_pivot

count                                 
age_frame             до 34 от 34 до 43 от 43 до 54 свыше 54
family_status                                               
в разводе               177         306         379      333
не женат / не замужем  1249         561         523      477
вдовец / вдова           13          64         204      678
гражданский брак       1123        1117        1040      871
женат / замужем        2804        3317        3302     2916

Данные показывают, что среди клиентов банка с семейными статусами **вдовец / вдова** и **в разводе** высокая доля тех, кому от 43 до 54 и тех, кому свыше 54, в тоже время среди клиентов банка с семейными статусами **не женат / не замужем** и **гражданский брак** доля тех, кому до 34 больше, чем остальных клиентов.

Далее проверим средний уровень дохода среди разных возрастных категорий клиентов банка, сгруппированные по семейным статусам.

In [59]:
data_pivot = df_with_family_status.pivot_table(index=['family_status', 'age_frame'], values='monthly_income', aggfunc=['mean'])
data_pivot = data_pivot.sort_values(by=('mean', 'monthly_income'), ascending=False)
data_pivot

mean
                                  monthly_income
family_status         age_frame                 
не женат / не замужем от 43 до 54  180512.692161
в разводе             от 34 до 43  180423.934641
женат / замужем       от 34 до 43  177032.721435
гражданский брак      от 34 до 43  174591.604297
не женат / не замужем от 34 до 43  173257.850267
женат / замужем       от 43 до 54  172754.919140
гражданский брак      от 43 до 54  171623.405769
вдовец / вдова        до 34        168119.384615
в разводе             от 43 до 54  166033.820580
женат / замужем       до 34        164994.323823
в разводе             до 34        163990.638418
гражданский брак      до 34        163412.720392
не женат / не замужем до 34        162758.728583
вдовец / вдова        от 34 до 43  160506.500000
в разводе             свыше 54     159885.210210
не женат / не замужем свыше 54     151958.283019
женат / замужем       свыше 54     150159.215706
вдовец / вдова        от 43 до 54  147723.269608
гражданский брак      свыше 54     145877.464983
вдовец / вдова        свыше 54     138865.525074

Данные показывают, что невозможно однозначно утверждать, что низкая подверженность к риску возникновения задолженности клиентов с семейными статусами **вдовец / вдова** и **в разводе** обусловлено высоким уровнем их дохода. Для точного ответа необходимы данные с суммой полученного кредита клиентами и сроками, под которые были взяты кредиты, но, к сожалению, такая информация отсутствует. 

### Вывод 2:

Клиенты с семейными статусами **вдовец / вдова** и **в разводе** менее всего подвержены риску возникновения задолженности. 
Больше всех к риску возникновения задолженности подвержены клиенты с семейными статусами **не женат / не замужем** и **гражданский брак**. Возможно, это обусловлено более осознанным подходом к кредиту.

Влияние дохода на погашение кредита является отдельной гипотезой, которую мы также проверим ниже.

[В начало](#intro).

### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [60]:
data_pivot = df.pivot_table(index=['monthly_income_category'], values='debt', aggfunc=['sum', 'count', make_proportion])
data_pivot = data_pivot.sort_values(by=('make_proportion', 'debt'))
data_pivot

sum  count make_proportion
                         debt   debt            debt
monthly_income_category                             
D                          21    350            6.0%
B                         356   5042           7.06%
A                           2     25            8.0%
C                        1360  16015           8.49%
E                           2     22           9.09%

### Вывод 3:

Данные показывают неоднозначный результат. Самые обеспеченные и менее всего обеспеченные клиенты практически в одинаковой степени неохотно берут кредит. Кредит очень охотно берет условный средний класс с уровнем дохода от 50 000 до 200 000, уровень риска по среднему классу находится чуть выше среднего. Самый больший риск характерен для менее всего обеспеченных клиентов банка, а вот меньше всего риск характерен не для самых обеспеченных клиентов банка. Менее всего подвержены риску кредиты, выданные клиентам с уровнем ежемесячного дохода до 50 000 и до 1 000 000.

[В начало](#intro).

### Вопрос 4: Как разные цели кредита влияют на его возврат в срок? 

In [61]:
data_pivot = df.pivot_table(index=['purpose_category'], values='debt', aggfunc=['sum', 'count', make_proportion])
data_pivot = data_pivot.sort_values(by=('make_proportion', 'debt'))
data_pivot

sum  count make_proportion
                         debt   debt            debt
purpose_category                                    
операции с недвижимостью  782  10811           7.23%
проведение свадьбы        186   2324            8.0%
получение образования     370   4013           9.22%
операции с автомобилем    403   4306           9.36%

### Вывод 4:

Данные показывают, что наименее рискованными являются кредиты, выданные на операции с недвижимостью и проведение свадьбы, в то время как наиболее рискованными выступают кредиты, выданные для получения образования и на операции с автомобилем. 

[В начало](#intro).

## Общий вывод:


Исследование показало портрет потенциального идеального клиента: человек состоящий/состоявший в официальных отношениях, без детей, с доходом выше среднего, цель получения кредита у которого недвижимость.

Отвечая на первый вопрос, одназначно можно утвреждать, что наличие детей у клиентов повышает риск возникновения задолженности, чем меньше детей, тем ниже риск просрочки по кредиту.

|Количество детей|Уровень риска | 
|:---------------|:------------:|          
|нет детей       |7.54%         | 
|много детей     |8.16%         | 
|1 ребенок       |9.17%         | 
|2 ребенка       |9.49%         |

---------------------------------

По второму вопросу иследование показало, что меньше всего риск просрочки по кредиту характерен для клиентов, состоящих/состоявшие в официальных отношениях, в то время как клиенты с гражданским браком и не состоящие в официальных отношениях больше всего подвержены риску возникновения задолженности по кредиту.


|Семейный статус       |Уровень риска | 
|:---------------------|:------------:|          
|вдовец / вдова        |6.57%         | 
|в разводе             |7.11%         | 
|женат / замужем       |7.55%         | 
|гражданский брак      |9.35%         | 
|не женат / не замужем |9.75%         |

---------------------------------------

Наибольший риск возникновения задолженности характерен для среднего класса, представителей среднего класса насчитывается большинство, в то время как самые обеспеченные и менее всего обеспеченные клиенты практически в одинаковой степени неохотно берут кредит, что же касается риска возникновения задолженности по низшему и высшему классу, то он достаточно высокий в обоих случаях. Иначе говоря, зависимость между уровнем дохода и погашением кредита в срок однозначно существует.

|Категория дохода|Уровень риска | 
|:---------------|:------------:|          
|D               |7.54%         | 
|B               |8.16%         | 
|A               |9.17%         | 
|C               |9.49%         | 
|E               |9.49%         | 

---------------------------------

Исследование показало, что разные цели кредита влияет на уровень риска возникновения задолженности, согласно данным исследования наименнее рискованным явлются с операции с недвижимостью, а наиболее рискованным операции с движемым имуществом.


|Категория дохода         |Уровень риска | 
|:------------------------|:------------:|          
|операции с недвижимостью |7.23%         | 
|проведение свадьбы       |8.0%          | 
|получение образования    |9.22%         | 
|операции с автомобилем   |9.36%         | 


[В начало](#intro).

## Рекомендации по повышению качества данных

В ходе анализа информации в датафрейме было выявлено достаточно много аномалий, что влияют на качество данных, а значит на точность исследования. В связи с этим по каждой выявленной аномалии предложим меры их по устренению:

- Для того чтобы в столбце с возрастом клиентов банка не встречались нулевые или отрицательные значения предлагается установить регулярные выражения на поле ввода возраста в случае, если заявка заполнется клиентом онлайн на сайте банка без помощи специлистов и если информация с заявки без предварительной обработки попадает в базу данных банка. Каким может быть регулярное вражение: ограничение на ввод нулевых или отрицательных значений, при вводе подобных значений клиент получит сообщение "возраст не может быть нулевым или отрицательным, пожалуйста введите свой возраст корректно". Тоже самое касается работы специалистов банка. Однако, если причиной появления нулей является не человеческий фактор, то следует выяснить причину появления подобного рода недостоверных данных при помощи команды разработчиков.

- Частота отрицательных значений в столбце с информацией о стажах работы клиентов, а также наличие сильно завышенных данных, снижают вероятность человеческого фактора как первопричины появления таких данных, следует обратиться к команде разработки.

- Низкая частота отрицательных и завышенных значений для столбца количеством детей у клиентов банка повышает вероятность человеческого фактора как первопричины появления таких амномалий в данных, по аналогии со столбцом с возрастом клиентов установить регулярные выражения на поле ввода количества детей.

- Для столбца с информацией об уровнях образования, а также для столбца с целями получения кредита и для любых категориальных данных установить выпадающие списки из строго ограниченного выбора значений, таким образом качество и точность данных в этих столбцах в разы увеличатся.

[В начало](#intro).